src/masternode/node.h

```diff
@@ -63,8 +60,38 @@ class CActiveMasternodeManager final : public CValidationInterface
 
     static bool IsValidNetAddr(const CService& addrIn);
 
+    template <template <typename> class EncryptedObj, typename Obj>
+    [[nodiscard]] bool Decrypt(const EncryptedObj<Obj>& obj, size_t idx, Obj& ret_obj, int version) const
+        LOCKS_EXCLUDED(cs);
+    [[nodiscard]] CBLSSignature Sign(const uint256& hash) const LOCKS_EXCLUDED(cs);
+    [[nodiscard]] CBLSSignature Sign(const uint256& hash, const bool is_legacy) const LOCKS_EXCLUDED(cs);
+
+    [[nodiscard]] const COutPoint& GetOutPoint() const EXCLUSIVE_LOCKS_REQUIRED(cs)
+    {
+        return m_info.outpoint;
+    }
+
+    [[nodiscard]] const uint256& GetProTxHash() const EXCLUSIVE_LOCKS_REQUIRED(cs)
+    {
+        return m_info.proTxHash;
+    }
+
+    [[nodiscard]] CBLSPublicKey GetPubKey() const EXCLUSIVE_LOCKS_REQUIRED(cs);
```

Comment:
We should lock internally if we are returning a copy; an external lock should only be required for a reference

Reply:
We are locking externally as opposed to internally to minimize the amount of code modification (existing code that read with `activeMasternodeInfo` used to hold the lock before attempting to read from it) and reducing the amount of lock-unlock events (functions that tend to read one value tend to read more).

Internal locking was used for functions introduced in this PR like `Decrypt` and `Sign`, because its usage doesn't usually involve accessing other members (where it might be reasonable to refactor the code to have all accesses under one locked scope), the exception to that rule is `CMNAuth::PushMNAUTH`, which I had to refactor so that it fetches values under one externally locked scope and signs outside, with locking done internally.

Reply:
Hmm.. Maybe let's add a TODO then? I do think we should only lock externally if we are getting a ref (or something else lifetime bound)

Reply:
Done